File: backend/src/controllers/report.controller.ts
import { Request, Response } from 'express';
import { PrismaClient } from '@prisma/client';
import { Role } from '../utils/constants';
import { TaskService } from '../services/task.service';

const prisma = new PrismaClient();

export const ReportController = {
  getHistory: async (req: Request, res: Response): Promise<void> => {
    try {
      const user = (req as any).user;
      if (!user) {
        res.status(401).json({ error: 'Unauthorized' });
        return;
      }

      // We'll fetch StatusHistory with its related Task and User objects
      // to provide a rich report
      let historyQuery: any = {
        include: {
          user: { select: { userId: true, name: true, username: true, role: true } },
          task: {
            include: {
              assignee: { select: { userId: true, name: true, username: true, role: true } },
              author: { select: { userId: true, name: true, username: true } },
              module: { select: { moduleName: true } },
              team: { select: { name: true } }
            }
          }
        },
        orderBy: {
          createdAt: 'desc'
        }
      };

      // RBAC filtering logic
      if (user.role === Role.SUPER_ADMIN || user.role === Role.ADMIN) {
        // Admins see everything. No extra where clause needed.
      } else if (user.role === Role.MANAGER) {
        // Managers see tasks authored by them, assigned to them, or assigned to their subordinates
        const subordinates = await prisma.user.findMany({
          where: { userAssignments: { some: { managerId: user.userId } } },
          select: { userId: true }
        });
        const subordinateIds = subordinates.map(sub => sub.userId);
        const allowedUserIds = [user.userId, ...subordinateIds];

        historyQuery.where = {
          OR: [
            { changedById: { in: allowedUserIds } },
            { task: { assigneeId: { in: allowedUserIds } } },
            { task: { authorId: { in: allowedUserIds } } }
          ]
        };
      } else {
        // Resources/Developers/Deployment see ONLY their own history
        historyQuery.where = {
          OR: [
            { changedById: user.userId },
            { task: { assigneeId: user.userId } },
            { task: { authorId: user.userId } }
          ]
        };
      }

      // Execute query
      const statusHistory = await prisma.statusHistory.findMany(historyQuery);
      
      const auditLogQuery: any = {
        include: {
          actor: { select: { userId: true, name: true, username: true, role: true } },
          task: {
            select: {
              id: true,
              title: true,
              status: true,
              lifecycleStatus: true,
              module: { select: { moduleName: true } },
            },
          },
        },
        orderBy: {
          changedAt: 'desc'
        }
      };

      if (user.role === Role.SUPER_ADMIN || user.role === Role.ADMIN) {
        // all
      } else if (user.role === Role.MANAGER) {
        const subordinates = await prisma.user.findMany({
          where: { userAssignments: { some: { managerId: user.userId } } },
          select: { userId: true }
        });
        const allowedUserIds = [user.userId, ...subordinates.map(s => s.userId)];
        auditLogQuery.where = { changedBy: { in: allowedUserIds } };
      } else {
        auditLogQuery.where = { changedBy: user.userId };
      }

      const auditLogs = await prisma.auditLog.findMany(auditLogQuery);

      res.status(200).json({
        success: true,
        data: {
          statusHistory,
          auditLogs
        }
      });
    } catch (error: any) {
      console.error("Error fetching report history:", error);
      res.status(500).json({ error: error.message });
    }
  },
  getReportData: async (req: Request, res: Response): Promise<void> => {
    try {
      const user = (req as any).user;
      if (!user) {
        res.status(401).json({ error: 'Unauthorized' });
        return;
      }

      if (user.role === Role.DEVELOPER || user.role === Role.VERIFIER) {
        res.status(403).json({ error: 'Forbidden: You do not have access to reports.' });
        return;
      }

      const { view, startDate, endDate, moduleId, clientId, managerId, developerId, verifierId, status } = req.query;

      // Base query scoped by role
      let whereClause: any = {
        lifecycleStatus: { lt: 100 }, // exclude soft-deleted
      };

      if (user.role === Role.CLIENT) {
        whereClause = {
          ...whereClause,
          OR: [{ authorId: user.userId }, { clientId: user.userId }],
        };
      } else if (user.role === Role.MANAGER) {
        const managedUserRelations = await prisma.userManager.findMany({
          where: { managerId: user.userId },
          select: { userId: true }
        });
        const teamUserIds = managedUserRelations.map(r => r.userId);
        whereClause = {
          ...whereClause,
          OR: [
            { authorId: user.userId },
            { managers: { some: { userId: user.userId } } },
            { developers: { some: { userId: { in: teamUserIds } } } },
            { verifiers: { some: { userId: { in: teamUserIds } } } }
          ],
        };
      }

      // Preset / custom date bounds on createdAt
      if (view === 'weekly') {
        const last7Days = new Date();
        last7Days.setDate(last7Days.getDate() - 7);
        whereClause.createdAt = { gte: last7Days };
      } else if (view === 'monthly') {
        const last30Days = new Date();
        last30Days.setDate(last30Days.getDate() - 30);
        whereClause.createdAt = { gte: last30Days };
      } else if (view === 'custom' && startDate && endDate) {
        whereClause.createdAt = {
          gte: new Date(startDate as string),
          lte: new Date(endDate as string),
        };
      }

      // Query filters
      if (moduleId) {
        whereClause.moduleId = moduleId as string;
      }
      if (clientId) {
        whereClause.clientId = clientId as string;
      }
      if (managerId) {
        whereClause.managers = { some: { userId: managerId as string } };
      }
      if (developerId) {
        whereClause.developers = { some: { userId: developerId as string } };
      }
      if (verifierId) {
        whereClause.verifiers = { some: { userId: verifierId as string } };
      }
      if (status) {
        whereClause.status = status as string;
      }

      // Execute query including all relations
      const tasks = await prisma.task.findMany({
        where: whereClause,
        include: {
          author: true,
          client: true,
          managers: true,
          developers: true,
          verifiers: true,
          team: true,
          module: true,
          statusHistory: { orderBy: { createdAt: "asc" } },
          comments: { include: { user: true } },
          attachments: true,
        },
        orderBy: { createdAt: "desc" },
      });

      // Resolve TaskService actor normalization to populate history acting users
      const normalized = tasks.map(t => TaskService.normalizeTask(t));
      const resolvedTasks = await TaskService.resolveTasksActors(normalized, tasks);

      res.status(200).json({
        success: true,
        data: resolvedTasks,
      });
    } catch (error: any) {
      console.error("Error fetching report data:", error);
      res.status(500).json({ error: error.message });
    }
  }
};

File: backend-spring/src/main/java/com/patchflow/controller/ReportController.java
package com.patchflow.controller;

import com.patchflow.config.Auth;
import com.patchflow.entity.*;
import com.patchflow.repository.*;
import com.patchflow.service.TaskService;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/reports")
@RequiredArgsConstructor
public class ReportController {

    private final StatusHistoryRepository statusHistoryRepository;
    private final AuditLogRepository auditLogRepository;
    private final TaskRepository taskRepository;
    private final UserManagerRepository userManagerRepository;
    private final TaskService taskService;

    @GetMapping("/history")
    @Transactional(readOnly = true)
    public ResponseEntity<?> getHistory(HttpServletRequest req) {
        User user = Auth.require(req);
        String role = user.getRole();
        String userId = user.getUserId();

        List<StatusHistory> statusHistory;
        List<AuditLog> auditLogs;

        if ("SUPER_ADMIN".equals(role) || "ADMIN".equals(role)) {
            statusHistory = statusHistoryRepository.findAll(
                    (root, query, cb) -> { query.orderBy(cb.desc(root.get("createdAt"))); return cb.conjunction(); }
            );
            auditLogs = auditLogRepository.findAll(
                    (root, query, cb) -> { query.orderBy(cb.desc(root.get("changedAt"))); return cb.conjunction(); }
            );
        } else if ("MANAGER".equals(role)) {
            List<String> subordinateIds = userManagerRepository.findByManagerId(userId)
                    .stream().map(UserManager::getUserId).collect(Collectors.toList());
            List<String> allowedIds = new ArrayList<>(subordinateIds);
            allowedIds.add(userId);

            statusHistory = statusHistoryRepository.findAll(
                    (root, query, cb) -> {
                        query.orderBy(cb.desc(root.get("createdAt")));
                        return cb.in(root.get("changedById")).value(allowedIds);
                    }
            );
            auditLogs = auditLogRepository.findAll(
                    (root, query, cb) -> {
                        query.orderBy(cb.desc(root.get("changedAt")));
                        return cb.in(root.get("changedBy")).value(allowedIds);
                    }
            );
        } else {
            statusHistory = statusHistoryRepository.findAll(
                    (root, query, cb) -> {
                        query.orderBy(cb.desc(root.get("createdAt")));
                        return cb.equal(root.get("changedById"), userId);
                    }
            );
            auditLogs = auditLogRepository.findAll(
                    (root, query, cb) -> {
                        query.orderBy(cb.desc(root.get("changedAt")));
                        return cb.equal(root.get("changedBy"), userId);
                    }
            );
        }

        return ResponseEntity.ok(Map.of("success", true, "data",
                Map.of("statusHistory", statusHistory, "auditLogs", auditLogs)));
    }

    @GetMapping("/data")
    @Transactional(readOnly = true)
    public ResponseEntity<?> getReportData(
            @RequestParam(required = false) String view,
            @RequestParam(required = false) String startDate,
            @RequestParam(required = false) String endDate,
            @RequestParam(required = false) String moduleId,
            @RequestParam(required = false) String clientId,
            @RequestParam(required = false) String managerId,
            @RequestParam(required = false) String developerId,
            @RequestParam(required = false) String verifierId,
            @RequestParam(required = false) String status,
            HttpServletRequest req) {

        User user = Auth.require(req);
        String role = user.getRole();
        String userId = user.getUserId();

        if ("DEVELOPER".equals(role) || "VERIFIER".equals(role)) {
            return ResponseEntity.status(403).body(Map.of("error", "Forbidden: You do not have access to reports."));
        }

        // Fetch all tasks and apply in-memory filtering (mirrors Prisma where clauses)
        List<Task> allTasks = taskRepository.findAll();
        List<Task> tasks = allTasks.stream()
                .filter(t -> t.getLifecycleStatus() < 100)
                .filter(t -> {
                    if ("CLIENT".equals(role)) return userId.equals(t.getAuthorId()) || userId.equals(t.getClientId());
                    if ("MANAGER".equals(role)) {
                        List<String> teamIds = userManagerRepository.findByManagerId(userId).stream().map(UserManager::getUserId).toList();
                        return t.getManagers().stream().anyMatch(m -> m.getUserId().equals(userId)) ||
                               userId.equals(t.getAuthorId()) ||
                               t.getDevelopers().stream().anyMatch(d -> teamIds.contains(d.getUserId())) ||
                               t.getVerifiers().stream().anyMatch(v -> teamIds.contains(v.getUserId()));
                    }
                    return true;
                })
                .filter(t -> {
                    if ("weekly".equals(view))  return t.getCreatedAt().isAfter(Instant.now().minusSeconds(7 * 86400));
                    if ("monthly".equals(view)) return t.getCreatedAt().isAfter(Instant.now().minusSeconds(30 * 86400));
                    if ("custom".equals(view) && startDate != null && endDate != null) {
                        Instant s = taskService.parseInstant(startDate, null);
                        Instant e = taskService.parseInstant(endDate, null);
                        if (s != null && e != null) {
                            return !t.getCreatedAt().isBefore(s) && !t.getCreatedAt().isAfter(e);
                        }
                    }
                    return true;
                })
                .filter(t -> moduleId == null   || moduleId.equals(t.getModuleId()))
                .filter(t -> clientId == null   || clientId.equals(t.getClientId()))
                .filter(t -> managerId == null  || t.getManagers().stream().anyMatch(m -> m.getUserId().equals(managerId)))
                .filter(t -> developerId == null || t.getDevelopers().stream().anyMatch(d -> d.getUserId().equals(developerId)))
                .filter(t -> verifierId == null  || t.getVerifiers().stream().anyMatch(v -> v.getUserId().equals(verifierId)))
                .filter(t -> status == null || status.equals(t.getStatus()))
                .sorted(Comparator.comparing(Task::getCreatedAt).reversed())
                .collect(Collectors.toList());

        List<Map<String, Object>> normalized = tasks.stream().map(taskService::normalizeTask).collect(Collectors.toList());
        return ResponseEntity.ok(Map.of("success", true, "data", normalized));
    }
}
